File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    answer = 0
    MOD = 1000000007
    
    p = set()
    for i,j in puddles:
        p.add((j-1, i-1))
    
    DP = [[0 for _ in range(m)] for _ in range(n)]
    for i in range(n):
        if (i,0) not in p:
            DP[i][0] = 1
        else:
            break
    for j in range(m):
        if (0,j) not in p:
            DP[0][j] = 1
        else:
            break
    
    for i in range(1, n):
        for j in range(1, m):
            if (i, j) in p:
                continue
            DP[i][j] = (DP[i-1][j] + DP[i][j-1]) % MOD

    
    return (DP[n-1][m-1])%MOD

File: 프로그래머스/3/42898. 등굣길/등굣길.js
function solution(N, M, puddles) {
    let answer = 0;
    const MOD = 1_000_000_007;
    const board = Array.from({length: N}, () => Array(M).fill(0));
    for (const [x, y] of puddles) {
        board[x-1][y-1] = -1;
    }
    
    board[0][0] = 1;
    const STEPS = [[0, 1], [1, 0]];
    for (let i=0; i<N; ++i) {
        for (let j=0; j<M; ++j) {
            for (let k=0; k<2; ++k) {
                const ni = i+STEPS[k][0];
                const nj = j+STEPS[k][1];
                if (ni>=N || nj >=M || board[i][j] === -1 || board[ni][nj] === -1) continue;
                board[ni][nj] += board[i][j];
                board[ni][nj] %= MOD;
            }
        }
    }
    
    return board[N-1][M-1];
}